### ISSUE TITLE
How do providers that use "no_cache" free their dispatch tables?

### ISSUE THREAD

**Issue by mattcaswell** [openssl/openssl#11771]
We have the capability for providers to specify that dispatch tables returned from their query function should not be cached (via the "no_cache" parameter). Currently no provider makes use of this.

All our providers at the moment use compiled-in "const" structures to hold their dispatch tables - so they never need to be freed. However a provider that needs "no_cache" presumably is doing so because those dispatch tables might change dynamically over time. In that case every time the query function is called presumably the provider might need to create a new set of dispatch tables in heap memory. In this scenario how does a provider free that memory up later?

It would be really good to have a test for a "no_cache" provider to test that this feature actually works.

**Comment by levitte** [comment]
I assume that said freeing would happen in the teardown call.

**Comment by levitte** [comment]
to be more precise, the spots where these dispatch table could be manipulated should be:

- provider init : first allocation
- operation query : reallocation if necessary (possibly free if a dispatch table "disappears")
- teardown : final free

**Comment by mattcaswell** [comment]
> I assume that said freeing would happen in the teardown call.

Well that's kind of ok - except that if there is no caching then the query function may be called many times. Each time it may need to create a new dispatch table - but presumably still keep the last one around in case its still being used by some thread. Hence there will be unbounded memory growth.

**Comment by mattcaswell** [comment]
Thread 1 and Thread 2 both call query. Thread 2 comes second and frees the dispatch table before thread 1 has finished using it.

**Comment by levitte** [comment]
Uhmm, no?  When libcrypto creates a method from a dispatch table, yeah sure, it uses that dispatch table...  but it doesn't save the pointer to the dispatch table anywhere, so it's not something that needs to be kept around for a huge amount of time at the very least.  We *might* have too have another look at locking, though.  Another alternative is to add an "unquery" function that libcrypto could call when it's done with a dispatch table.

**Comment by mattcaswell** [comment]
> We might have too have another look at locking, though.

Well, that's really the point I was trying to make. It doesn't matter if these things are short lived - its a non zero amount of time which is enough for two threads to trample on each other.

I have a tendency to prefer the "unquery" concept to avoid unnecessary locking. Otherwise, for a non-caching provider, I imagine locking could become a major bottleneck. The "unquery" function could be entirely optional. If the provider has one it gets called. Otherwise it doesn't.

**Comment by levitte** [comment]
> I have a tendency to prefer the "unquery" concept to avoid unnecessary locking.

I like your tendencies :wink:

So, er, PR welcome!  (or should I?)

**Comment by mattcaswell** [comment]
> So, er, PR welcome! (or should I?)

Who ever gets to it first. Not urgent, but should be done before beta 1 IMO.

**Comment by mattcaswell** [comment]
Adding to 3.0. beta 1 milestone

Ping @levitte!! Thoughts?

**Comment by levitte** [comment]
No more thoughts than above.
I have only a vague idea how to do this. It should be a fairly simple call when a function like EVP_CIPHER_free reaches zero on the reference count...

**Comment by mattcaswell** [comment]
> I have only a vague idea how to do this for the moment. It should be a fairly simple call when a function like EVP_CIPHER_free reaches zero on the reference count...

Isn't simply a case of calling "unquery" after this block:

https://github.com/openssl/openssl/blob/21e5be854deb65f54661c8231a9a30a453a173e0/crypto/core_algorithm.c#L59-L67

After that `map` seems to go out of scope and isn't saved. Unless that is `data->fn` assumes that the reference provided in `thismap` is stable and can be cached.

**Comment by mattcaswell** [comment]
Or, alternatively we put a lock around this section so that the provider can assume that query_operation will only ever be called once at a time for a given libctx (but that might have performance implications for no-cache providers).

**Comment by slontis** [comment]
@paulidale any thoughts on this?

**Comment by paulidale** [comment]
Rather than locking, I'd be more tempted to add a `provider_query_done()` or `provider_dispatch_free()` call to allow a provider to clean up the returned dispatch table.  For all existing providers this wouldn't be defined.

The same issue exists for the gettable/settable functions -- there is no reason these couldn't change dynamically.

**Comment by levitte** [comment]
> Isn't it simply a case of calling "unquery" after this block:

Yes, I think it is, and the "unquery" would then only be done if `no_cache` is 1.

**Comment by levitte** [comment]
> The same issue exists for the gettable/settable functions -- there is no reason these couldn't change dynamically.

Not sure I understand...  the functions themselves should stay in memory no matter what, shouldn't they? (unless someone unloads the module that has them, but I believe we have protected ourselves against that pretty well)
You might mean that the pointers to them in the OSSL_DISPATCH table might change, but that shouldn't matter, since we take copies of all interesting pointers when creating the methods.

**Comment by levitte** [comment]
I think there's a misunderstanding of how `query_operation` is assumed to work.  The idea is that, for a given provider and operation ID, `query_operation` only has one answer at any given point in time, and the `no_cache` parameter indicates to the caller if it's safe to assume if that OSSL_ALGORITHM is constant throughout the lifetime of the loaded provider, or if it may change from one `query_operation` to the next.

This isn't quite the same as viewing `query_operation` as a malloc(), since allocation may imply that there may exist multiple OSSL_ALGORITHM arrays for a given provider and operation ID at the same time.  That implication is a huge change of the design.

Another way to see things is that one could view the `query_operation` function as an implied lock of the OSSL_ALGORITHM array, to be unlocked by the `unquery_operation` function.  That would be more in line with the current design.
